mysql/datadog_checks/mysql/execution_plans.py

```diff
@@ -175,20 +164,156 @@ def _collect_execution_plans(self, db, tags, options):
+        preferred_tables = ['events_statements_history_long', 'events_statements_history', 'events_statements_current']
+        events_statements_table = options.get('events_statements_table', None)
+        if events_statements_table and events_statements_table not in DEFAULT_EVENTS_STATEMENTS_RATE_LIMITS:
+            self.log.warning("invalid events_statements_table: %s. must be one of %s", events_statements_table,
+                             ', '.join(DEFAULT_EVENTS_STATEMENTS_RATE_LIMITS.keys()))
+            events_statements_table = None
+        if events_statements_table:
+            preferred_tables = [events_statements_table]
+
+        # default time limit is small enough (1ms) so that the check will run only once
+        chosen_table = None
+        collect_exec_plans_time_limit = options.get('collect_exec_plans_time_limit', 1 / 1000)
+        collect_exec_plans_rate_limit = options.get('collect_exec_plans_rate_limit', -1)
+
+        for table in preferred_tables:
+            if table not in enabled_consumers:
+                if not auto_enable:
+                    self.log.debug("performance_schema consumer for table %s not enabled")
+                    continue
+                success = self._performance_schema_enable_consumer(db, table)
+                if not success:
+                    continue
+                self.log.debug("successfully enabled performance_schema consumer")
+
+            rows = self._get_events_statements_by_digest(db, table, 1)
+            if not rows:
+                self.log.debug("no statements found in %s", table)
+                continue
+
+            if collect_exec_plans_rate_limit < 0:
+                collect_exec_plans_rate_limit = DEFAULT_EVENTS_STATEMENTS_RATE_LIMITS[table]
+            if collect_exec_plans_time_limit < 1 and table != 'events_statements_history_long':
+                # all other tables require sampling multiple times during a single check run, so set the time limit
+                # to run for most of the check run, leaving one second free to ensure it doesn't go over
+                collect_exec_plans_time_limit = max(1, min_collection_interval - 1)
+
+            chosen_table = table
+            break
+
+        strategy = (
+            chosen_table,
+            collect_exec_plans_time_limit,
+            collect_exec_plans_rate_limit
+        )
+
+        if chosen_table:
+            # cache only successful strategies, 10 minute expiry
+            # 10 minutes is short enough that we'll reflect updates "relatively quickly"
+            # i.e., an aurora replica becomes a master (or vice versa)
+            self.log.debug("found plan collection strategy. chosen_table=%s, time_limit=%s, rate_limit=%s",
+                           chosen_table, collect_exec_plans_time_limit, collect_exec_plans_rate_limit)
+            self._expiring_cache.set("plan_collection_strategy", strategy, 10 * 60)
+        else:
+            self.log.warning(
```

Comment:
Since we don't cache the null state, we'll just keep trying every check run (which will be an error in query metrics). Is that what we want to do? We _might_ want to cache here for a bit, maybe a shorter period like 3 min. Do you agree?

Reply:
Discussed offline, going to leave the null state uncached for now.